flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/StopWithSavepointTerminationHandlerImplTest.java

```diff
@@ -0,0 +1,330 @@
+ */
+public class StopWithSavepointTerminationHandlerImplTest extends TestLogger {
+
+    @ClassRule public static final TemporaryFolder TEMPORARY_FOLDER = new TemporaryFolder();
+
+    private static final JobID JOB_ID = new JobID();
+
+    private final TestingCheckpointScheduling checkpointScheduling =
+            new TestingCheckpointScheduling(false);
+
+    private StopWithSavepointTerminationHandlerImpl createTestInstance(
+            Consumer<Throwable> handleGlobalFailureConsumer) {
+        // checkpointing should be always stopped before initiating stop-with-savepoint
+        checkpointScheduling.stopCheckpointScheduler();
+
+        final SchedulerNG scheduler =
+                TestingSchedulerNG.newBuilder()
+                        .setHandleGlobalFailureConsumer(handleGlobalFailureConsumer)
+                        .build();
+        return new StopWithSavepointTerminationHandlerImpl(
+                JOB_ID, scheduler, checkpointScheduling, log);
+    }
+
+    @Test
+    public void testHappyPathWithSavepointCreationBeforeSuccessfulTermination() throws Exception {
+        assertHappyPath(
+                (completedSavepoint,
+                        completedSavepointFuture,
+                        terminatedExecutionStates,
+                        executionsTerminatedFuture) -> {
+                    completedSavepointFuture.complete(completedSavepoint);
+                    executionsTerminatedFuture.complete(terminatedExecutionStates);
+                });
+    }
+
+    @Test
+    public void testHappyPathWithSavepointCreationAfterSuccessfulTermination() throws Exception {
+        assertHappyPath(
+                (completedSavepoint,
+                        completedSavepointFuture,
+                        terminatedExecutionStates,
+                        executionsTerminatedFuture) -> {
+                    executionsTerminatedFuture.complete(terminatedExecutionStates);
+                    completedSavepointFuture.complete(completedSavepoint);
+                });
+    }
+
+    @Test
+    public void testSavepointCreationFailureBeforeSuccessfulTermination() {
+        assertSavepointCreationFailure(
+                (expectedException, completedSavepointFuture, executionsTerminatedFuture) -> {
+                    completedSavepointFuture.completeExceptionally(expectedException);
+                    executionsTerminatedFuture.complete(
+                            Collections.singletonList(ExecutionState.FINISHED));
+                });
+    }
+
+    @Test
+    public void testSavepointCreationFailureAfterSuccessfulTermination() {
+        assertSavepointCreationFailure(
+                (expectedException, completedSavepointFuture, executionsTerminatedFuture) -> {
+                    executionsTerminatedFuture.complete(
+                            Collections.singletonList(ExecutionState.FINISHED));
+                    completedSavepointFuture.completeExceptionally(expectedException);
```

Comment:
If we make the ordering of signals the responsibility of a different component, then we wouldn't have this combinatorial explosion of test cases. Then we would have test cases for the different signal order completions and test cases for the `StopWithSavepointTerminationHandlerImpl` where we can concentrate on the handler specific functionality.